### PR TITLE
Add max buffer count config property for optimized repartitioning

### PR DESCRIPTION
**Pull request by yingsu00**
```
== RELEASE NOTES ==

General Changes
* Add config property driver.max-page-partitioning-buffer-count, to limit the total number of buffers per optimized repartitioning operator. The default value is 1,000,000.
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java [hunk 1]
```diff
@@ -367,7 +367,8 @@ public OperatorFactory duplicate()
         private final Closer blockLeaseCloser = Closer.create();
 
         // The ArrayAllocator for the buffers used in repartitioning, e.g. PartitionBuffer#serializedRowSizes, BlockEncodingBuffer#mappedPositions.
-        private final ArrayAllocator bufferAllocator = new SimpleArrayAllocator(10_000);
+        // Assuming there are 1000 columns, and each of them uses 3 buffers, and the number of partitions is 333, then we need 1000 * 3 * 333 about 1M buffers.
+        private final ArrayAllocator bufferAllocator = new SimpleArrayAllocator(1_000_000);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java:371]: This number is highly dependent on the environment and won't work for all setups. Let's make this limit configurable.

**Comment by yingsu00** [comment]
@mbasmanova  Hi Masha, I have added a new config property `driver.max-page-partitioning-buffer-count` and made the default value 1 million. Will you be able to take another look? Thank you!

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java [hunk 2]
```diff
@@ -60,6 +60,7 @@
 
     private DataSize sinkMaxBufferSize = new DataSize(32, Unit.MEGABYTE);
     private DataSize maxPagePartitioningBufferSize = new DataSize(32, Unit.MEGABYTE);
+    private Integer maxPagePartitioningBufferCount = 1_000_000;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:63]: Integer -> int

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java [hunk 3]
```diff
@@ -384,6 +385,19 @@ public TaskManagerConfig setMaxPagePartitioningBufferSize(DataSize size)
         return this;
     }
 
+    @NotNull
+    public Integer getMaxPagePartitioningBufferCount()
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:389]: - remove @NotNull
- Integer -> int

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java [hunk 4]
```diff
@@ -384,6 +385,19 @@ public TaskManagerConfig setMaxPagePartitioningBufferSize(DataSize size)
...
+    {
+        return maxPagePartitioningBufferCount;
+    }
+
+    @Config("driver.max-page-partitioning-buffer-count")
+    public TaskManagerConfig setMaxPagePartitioningBufferCount(Integer bufferCount)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:395]: Integer -> int

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java [hunk 5]
```diff
@@ -384,6 +385,19 @@ public TaskManagerConfig setMaxPagePartitioningBufferSize(DataSize size)
         return this;
     }
 
+    @NotNull
+    public Integer getMaxPagePartitioningBufferCount()
+    {
+        return maxPagePartitioningBufferCount;
+    }
+
+    @Config("driver.max-page-partitioning-buffer-count")
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:394]: Add @ConfigDescription

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 6]
```diff
@@ -310,6 +310,7 @@
     private final IndexJoinLookupStats indexJoinLookupStats;
     private final DataSize maxPartialAggregationMemorySize;
     private final DataSize maxPagePartitioningBufferSize;
+    private final Integer maxPagePartitioningBufferCount;
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:313]: Integer -> int

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java (hunk 5 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:394]: Sure. Added `@ConfigDescription("Maximum number of buffers used by repartitioning per driver")`

**Comment by yingsu00** [comment]
@mbasmanova Hi Masha, thanks very much for the review! I have addressed all your new comments

> @yingsu00 Looks good to me. Each operator creates its own ArrayAllocator. Would it be better to compute the limit dynamically? The formula might be `<num-columns> * <num-partitions> * 3`.

@mbasmanova It is possible to calculate the limit dynamically. But the bufferAllocator was created upon the operator creation, while the number of buffers can only be decided when the first page comes in.  The calculation can be done in initializeBlockEncodingBuffers(). I may need to think it over and implement it after my oncall finishes. It is additional work for each buffer. It is unlikely to regress the performance but I also need to make sure. Can we merge this PR to unblock the release for now?

**Comment by yingsu00** [comment]
> @mbasmanova Hi Masha, thanks very much for the review! I have addressed all your new comments
> 
> > @yingsu00 Looks good to me. Each operator creates its own ArrayAllocator. Would it be better to compute the limit dynamically? The formula might be `<num-columns> * <num-partitions> * 3`.
> 
> @mbasmanova It is possible to calculate the limit dynamically. But the bufferAllocator was created upon the operator creation, while the number of buffers can only be decided when the first page comes in. The calculation can be done in initializeBlockEncodingBuffers(). I may need to think it over and implement it after my oncall finishes. It is additional work for each buffer. It is unlikely to regress the performance but I also need to make sure. Can we merge this PR to unblock the release for now?

We do have the sourceTypes, however, for nested types like array, it's hard to get the number of buffers because it may be deeply nested and we may need to add repartitioning-specific logic to Type interface to get the number of buffers for each level.

File: presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java [hunk 7]
```diff
@@ -384,6 +385,19 @@ public TaskManagerConfig setMaxPagePartitioningBufferSize(DataSize size)
         return this;
     }
 
+    public int getMaxPagePartitioningBufferCount()
+    {
+        return maxPagePartitioningBufferCount;
+    }
+
+    @Config("driver.max-page-partitioning-buffer-count")
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:393]: `@Min(1)`

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/execution/TaskManagerConfig.java:394]: Thanks @tdcmeehan ! I updated the PR with the following
```
    @Min(1)
    public int getMaxPagePartitioningBufferCount()
    {
        return maxPagePartitioningBufferCount;
    }
```

**Comment by caithagoras** [comment]
@yingsu00 Just retried. It should fix itself.

**Comment by yingsu00** [comment]
> @yingsu00 Just retried. It should fix itself.

Build #267 failed again. It says
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project presto-facebook-verifier: Compilation failure
[ERROR] /tmp/workspace/presto-facebook-oss-integration-github/presto-facebook-verifier/src/main/java/com/facebook/presto/verifier/PrestoFacebookVerifyCommand.java:[43,7] error: PrestoFacebookVerifyCommand is not abstract and does not override abstract method getTablePropertyOverrides() in VerifyCommand
[ERROR] -> [Help 1]
```

**Comment by caithagoras** [comment]
@yingsu00 Oh, I think you need to rebase this PR.

**Comment by mbasmanova** [comment]
@yingsu00 

> We do have the sourceTypes, however, for nested types like array, it's hard to get the number of buffers because it may be deeply nested and we may need to add repartitioning-specific logic to Type interface to get the number of buffers for each level.

Got it. We don't need to do this right now, but in the future this could be done without "introducing repartitioning-specific logic to Type interface". Instead, we can use a visitor pattern to traverse the type hierarchy and introduce a visitor that calculates the number of buffers needed for repartitioning.

**Comment by yingsu00** [comment]
@mbasmanova Hi Masha, can we merge this PR now? The tests have passed. Thank you!

**Comment by yingsu00** [comment]
> use a visitor pattern to traverse the type hierarchy and introduce a visitor that calculates the number of buffers needed for repartitioning.

Thanks for the suggestion @mbasmanova. I created issue https://github.com/prestodb/presto/issues/14364 for this.

**Pull request merged**